src/TreeNode.ts

```diff
@@ -0,0 +1,201 @@
+            inputWeight: vertexArray,
+            hasVertexError: isError,
+            hasWeightError: this.checkWeight(isError, getNonNeighbours(label, node.graph), vertexArray),
+            position: {...node.position},
+            children: []
+        }));
+    }
+
+    public removeLeaf(nodeId: number): void {
+        const parentNode = this.deepNodeSearch((n) => n.children.some(e => e.id === nodeId));
+        if (parentNode === void 0) { throw Error("Root cannot be removed."); }
+        parentNode.children = parentNode.children.filter(e => e.id !== nodeId);
+    }
+
+    public getNodeById(id: number): TreeNode {
+        const node = this.deepNodeSearch((n) => n.id === id);
+        if (node === void 0) { throw Error("No node found by specified id."); }
+        return node;
+    }
+
+    public getNodeList(): TreeNode[] {
+        const childrenNodeList:TreeNode[] = this.children.map(e => e.getNodeList()).reduce((a,b) => a.concat(b), []);
+        const self: TreeNode = this;
+        return [self].concat(childrenNodeList).sort((a,b) => a.getId() - b.getId());
+    }
+
+    public deepNodeSearch(predicate: (node: TreeNode) => boolean): TreeNode | void {
+        if (predicate(this)) { return this; }
+        if (this.children.length > 0) {
+            return this.children.map(e => e.deepNodeSearch(predicate)).find(e => e !== void 0);
+        }
+        return void 0;
+    }
+
+    private checkWeight(isError: boolean, real: string[], input: string[]) {
+        if (isError) {
+            {
+                Notifier.send({
+                    fee: 0,
+                    datetime: new Date().getTime(),
+                    message: "Так как была введена ошибочная вершина, ее вес так же считается ошибочным",
+                    variantId: 0,
+                    isTaskFinished: false,
+                });
+                return true;
+            }
+        }
+        else {
+            // проверяем, что введенное множество вершин соответствует рассчитанной неокрестности вершины
+            // если несоответствие, то записываем ошибку
+            if (real.sort().toString() === input.sort().toString()) {
+                return false
+            }
+            else {
+                Notifier.send({
+                    fee: 0,
+                    datetime: new Date().getTime(),
+                    message: "Введеное множество вершин неверно",
+                    variantId: 0,
+                    isTaskFinished: false,
+                });
+                return true;
+            }
+        }
+    }
+
+    private checkVertex(parentNode: TreeNode, label: string) {
+        // проверка, есть ли введенная вершина в графе
+        if (this.graph.vertices.map((e: IVertex) => e.name).includes(label)) {
+            // проверка, есть ли введенная вершина в графе, соответствующем узлу-родителю
+            if (parentNode.graph.vertices.map(e => e.name).includes(label)) {
```

Comment:
Почему строчкой выше есть тип, а тут нету?
```suggestion
            if (parentNode.graph.vertices.map((e: IVertex) => e.name).includes(label)) {
```

Reply:
Исправлено